62.CAD/vertex_shader.hlsl

```diff
@@ -2,18 +2,213 @@
 
 #include "common.hlsl"
 
+float2 Tangent(float2 p0, float2 p1, float2 p2, float t)
+{
+    float2 tangent = 2.0 * (1.0 - t) * (p1 - p0) + 2.0 * t * (p2 - p1);
+    return tangent;
+}
+
+float TforTan(float2 p0, float2 p1, float2 p2, float2 DirNormalized)
+{
+    float a = -2 * (p1.x - p0.x) + 2 * (p2.x - p1.x);
+        float b = 2 * (p1.x - p0.x);
+        float c = -2 * (p1.y - p0.y) + 2 * (p2.y - p1.y);
+        float d = 2 * (p1.y - p0.y);
+        float e = DirNormalized.x;
+        float f = DirNormalized.y;
+
+        float num = float(d * e - b * f);
+        float den = float(a * f - c * e);
+        float t;
+    if (den == 0) return 0;
+    else t = num / den;
+    return t;
+}
+
+float2 QuadraticBezier(float2 p0, float2 p1, float2 p2, float t)
+{
+    float tt = t * t;
+    float oneMinusT = 1.0 - t;
+    float oneMinusTT = oneMinusT * oneMinusT;
+    float2 position = p0 * oneMinusTT + 2.0 * p1 * oneMinusT * t + p2 * tt;
+    return position;
+}
+
+
+float2 CalculateIntersection(float2 line1Start, float2 line2Start, float2 line1End, float2 line2End)
+{
+    float2 intersectionPoint;
+
+    float2 line1Direction = normalize(line1End - line1Start);
+    float2 line2Direction = normalize(line2End - line2Start);
+
+    float line1Denominator = line2Direction.y * line1Direction.x - line2Direction.x * line1Direction.y;
+
+    if (abs(line1Denominator) < 0.0001)
+    {
+        // Lines are parallel, return a default value or handle accordingly.
+        intersectionPoint = float2(0, 0); // Default value, adjust as needed.
+    }
+    else
+    {
+        float2 line2StartToLine1Start = line1Start - line2Start;
+        float line2Numerator = line2Direction.x * line2StartToLine1Start.y - line2Direction.y * line2StartToLine1Start.x;
+
+        float t = line2Numerator / line1Denominator;
+        intersectionPoint = line1Start + t * line1Direction;
+    }
+
+    return intersectionPoint;
+}
+
+float3 normalize(float3 A) {
+    return A / sqrt(dot(A, A));
+}
+
+
+bool estimateTransformation(float2 p01, float2 p11, float2 p21, out float2 translation, out float2x2 rotation)
+{
+
+
+    float2 p1 = p11 - p01;
+    float2 p2 = p21 - p01;
+
+
+
+    float2 a = p2 - 2.0 * p1;
+    float2 b = 2.0 * p1;
+
```

Comment:
what's with all these empty spaces? 
add 2-3 comments in this function describing what you're trying to compute and the genral steps you're taking and what the variables mean